### PR TITLE
MIR-Validate StorageLive.

### PR DESCRIPTION
**Pull request by cjgillot**
`StorageLive` statements on a local which already has storage is banned by miri.

This check is easy enough, and can detect bugs in MIR opts.

### CODE REVIEW

File: compiler/rustc_const_eval/src/transform/validate.rs [hunk 1]
```diff
@@ -755,8 +755,19 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                     self.fail(location, format!("explicit `{:?}` is forbidden", kind));
                 }
             }
-            StatementKind::StorageLive(..)
-            | StatementKind::StorageDead(..)
+            StatementKind::StorageLive(local) => {
+                if self.reachable_blocks.contains(location.block) {
```
Comment (tmiasko) [inline compiler/rustc_const_eval/src/transform/validate.rs:759]: Can you reference the corresponding comment from visit_local method here? (Conceptually this shouldn't be part of a validator, but rather a separate optional lint pass, but as long as that comment is true that seems reasonable).

**Comment by wesleywiser** [comment]
I assume we want a perf run:

@rust-timer build 47a7405fd6f67fc959d6d12979f06fb1b5dcafac

**Review by wesleywiser — APPROVED** [review]
r=me with @tmiasko's suggestion applied

**Comment by pnkfelix** [comment]
* #108175 regressed helloworld check by 0.6% in three incremental scenarios. 
* note though that these regressions did not show up in the timer run before the PR landed.
* marking as triaged.

@rustbot label: perf-regression-triaged

**Pull request merged**